### PR TITLE
malicious network tests

### PR DESCRIPTION
**Pull request by bmcase**
add ability for client (RC) to request that the servers run a malicious IPA query and tests for this.

### CODE REVIEW

**Comment by bmcase** [comment]
@akoshelev can you help review this?   Not sure what is going on with `runner/ipa.rs` tests -- there are warnings to remove the imports but if they are removed their are compile errors.  Also some branch conflict with this part.

File: src/net/http_serde.rs [hunk 1]
```diff
@@ -135,7 +135,29 @@ pub mod query {
...
+                        max_breakdown_key: u32,
+                        attribution_window_seconds: Option<NonZeroU32>,
+                        num_multi_bits: u32,
+                    }
+                    let Query(IPAQueryConfigParam {
+                        per_user_credit_cap,
+                        max_breakdown_key,
+                        attribution_window_seconds,
+                        num_multi_bits,
+                    }) = req.extract().await?;
```
Comment (richajaindce) [inline src/net/http_serde.rs:158]: Can this be combined like for semihonest?

File: src/helpers/transport/query.rs [hunk 2]
```diff
@@ -248,6 +251,6 @@ impl IpaQueryConfig {
 
 impl From<IpaQueryConfig> for QueryType {
     fn from(value: IpaQueryConfig) -> Self {
```
Comment (richajaindce) [inline src/helpers/transport/query.rs:253]: Where is this used? Dont we need it for malicious also?

File: src/query/executor.rs [hunk 3]
```diff
@@ -73,18 +73,36 @@ pub fn start_query(
         // Negotiate PRSS first
         let step = step::Descriptive::default().narrow(&config.query_type);
         let prss = negotiate_prss(&gateway, &step, &mut rng).await.unwrap();
-        let ctx = SemiHonestContext::new(&prss, &gateway);
+        //let ctx = SemiHonestContext::new(&prss, &gateway);
```
Comment (richajaindce) [inline src/query/executor.rs:76]: remove this

File: src/query/executor.rs [hunk 4]
```diff
@@ -73,18 +73,36 @@ pub fn start_query(
...
+                    .run(
+                        SemiHonestContext::new(&prss, &gateway),
+                        config.field_type,
+                        input,
+                    )
+                    .await
+            }
+            QueryType::MaliciousIpa(ipa_query_config) => {
+                IpaRunner(ipa_query_config)
+                    //.run(SemiHonestContext::new(&prss, &gateway), config.field_type, input)
```
Comment (richajaindce) [inline src/query/executor.rs:100]: remove the commented line

File: src/cli/playbook/mod.rs [hunk 5]
```diff
@@ -1,5 +1,5 @@
 mod input;
-mod ipa;
+pub mod ipa;
```
Comment (akoshelev) [inline src/cli/playbook/mod.rs:2]: I think we can keep it private and re-export the malicious function like this

```rust
pub use self::ipa::malicious;
```

or if we can make a single function like suggested above, we won't even need that

File: src/helpers/transport/query.rs (hunk 2 above)
Comment (akoshelev) [reply src/helpers/transport/query.rs:253]: good finding, I wasn't able to locate a single use of this conversion, so we can just try to delete it and see if any check fails. I spot-checked shuttle tests and benches and received no complains about removal

**Comment by andyleiserson** [comment]
This overlaps my upcoming changes for match key decryption, and probably also #664, so we'll need to pay attention to the sequencing.

File: src/net/http_serde.rs (hunk 1 above)
Comment (bmcase) [reply src/net/http_serde.rs:158]: I'm not sure I follow what you mean.  These appear to also be separate from semihonest in lines 125-136.  what am I missing?

File: src/cli/playbook/ipa.rs [hunk 6]
```diff
@@ -102,3 +102,91 @@ where
 
     breakdowns
 }
+
+/// Malicious IPA protocol.
+/// Returns aggregated values per breakdown key represented as index in the returned vector
+#[allow(clippy::missing_panics_doc)]
+pub async fn malicious<F, MK, BK>(
```
[the remark this answers is not in the document]
Comment (bmcase) [reply src/cli/playbook/ipa.rs:109]: yep, fixed this too

File: src/cli/playbook/mod.rs (hunk 5 above)
Comment (bmcase) [reply src/cli/playbook/mod.rs:2]: yeah, was able to just remove after the change above

**Comment by bmcase** [comment]
Other than the changes @richajaindce  suggested in `http_serde,rs`, I've incorporated this feedback.  cc @akoshelev

**Comment by bmcase** [comment]
okay, I compressed the matched code in `http_serde.rs` as @richajaindce  suggested.

File: src/net/http_serde.rs (hunk 1 above)
Comment (bmcase) [reply src/net/http_serde.rs]: okay, updated this to remove some of the duplicate code in the match.

**Comment by akoshelev** [comment]
@bmcase you need         #[allow(clippy::large_futures)] annotation inside the newly created test. I would suggest merging the latest changes from `main` and running `rustup update` - you would need the latest compiler version to reproduce the error. Andy already fixed the similar issue: https://github.com/private-attribution/ipa/blob/2371b14f9fc23f4787d22049f0756d1c6ae0cae6/src/query/runner/ipa.rs#L116

**Comment by bmcase** [comment]
okay merged the main branch into the PR and fixed the clippy errors.

File: src/cli/playbook/ipa.rs [hunk 7]
```diff
@@ -102,3 +102,91 @@ where
...
+//         // I think using u32 is wrong, we should move to u128
+//         let trigger_value = u32::try_from(row.trigger_value.as_u128()).unwrap();
+//         if breakdown_key >= breakdowns.len() {
+//             breakdowns.resize(breakdown_key + 1, 0);
+//             breakdowns[breakdown_key] += trigger_value
+//         }
+//     }
+//
+//     breakdowns
+// }
```
Comment (akoshelev) [inline src/cli/playbook/ipa.rs:192]:
```suggestion
```

File: src/net/http_serde.rs [hunk 8]
```diff
@@ -157,11 +171,27 @@ pub mod query {
             match self.query_type {
                 #[cfg(any(test, feature = "test-fixture", feature = "cli"))]
                 QueryType::TestMultiply => write!(f, "query_type={}", QueryType::TEST_MULTIPLY_STR),
-                QueryType::Ipa(config) => {
+                QueryType::SemiHonestIpa(config) => {
```
Comment (akoshelev) [inline src/net/http_serde.rs:174]: Maybe we can replace it with

```rust
                qt @ (QueryType::SemiHonestIpa(config) | QueryType::MaliciousIpa(config)) => {
                    write!(
                        f,
                        "query_type={qt}&per_user_credit_cap={}&max_breakdown_key={}&num_multi_bits={}",
                        config.per_user_credit_cap,
                        config.max_breakdown_key,
                        config.num_multi_bits,
                        qt=qt.as_ref(),
                    )?;

                    if let Some(window) = config.attribution_window_seconds {
                        write!(f, "&attribution_window_seconds={}", window.get())?;
                    }

                    Ok(())
                }
```

this covers both malicious and semi-honest case and does not require code repetition for attribution window

**Review by akoshelev — APPROVED** [review]

File: src/net/http_serde.rs (hunk 8 above)
Comment (bmcase) [reply src/net/http_serde.rs]: okay, added this

File: src/cli/playbook/ipa.rs (hunk 7 above)
Comment (bmcase) [reply src/cli/playbook/ipa.rs]: yep, forgot to delete that old code.

**Pull request merged**